File: src/game/training/trainingRank.ts
import type { TrainingKind } from '@/game/training/trainingTypes';

export type TrainingLetterRank = 'S' | 'A' | 'B' | 'C' | 'D' | 'E' | 'F';

const DEFAULT_RANK_THRESHOLDS: readonly { minScore: number; rank: TrainingLetterRank }[] = [
  { minScore: 60, rank: 'S' },
  { minScore: 50, rank: 'A' },
  { minScore: 40, rank: 'B' },
  { minScore: 30, rank: 'C' },
  { minScore: 20, rank: 'D' },
  { minScore: 10, rank: 'E' },
];

const SCALE_RANK_THRESHOLDS: readonly { minScore: number; rank: TrainingLetterRank }[] = [
  { minScore: 30, rank: 'S' },
  { minScore: 25, rank: 'A' },
  { minScore: 20, rank: 'B' },
  { minScore: 15, rank: 'C' },
  { minScore: 10, rank: 'D' },
  { minScore: 5, rank: 'E' },
];

const thresholdsForKind = (kind: TrainingKind): readonly { minScore: number; rank: TrainingLetterRank }[] => (
  kind === 'scale' ? SCALE_RANK_THRESHOLDS : DEFAULT_RANK_THRESHOLDS
);

export const scoreToTrainingRank = (score: number, kind: TrainingKind = 'chord'): TrainingLetterRank => {
  const normalized = Math.max(0, Math.floor(score));
  for (const entry of thresholdsForKind(kind)) {
    if (normalized >= entry.minScore) {
      return entry.rank;
    }
  }
  return 'F';
};

const RANK_ORDER: readonly TrainingLetterRank[] = ['F', 'E', 'D', 'C', 'B', 'A', 'S'];

export const meetsTrainingRankByLetter = (
  achievedRank: TrainingLetterRank,
  requiredRank: TrainingLetterRank,
): boolean => RANK_ORDER.indexOf(achievedRank) >= RANK_ORDER.indexOf(requiredRank);

export const meetsTrainingRankRequirement = (
  score: number,
  requiredRank: TrainingLetterRank,
  kind: TrainingKind = 'chord',
): boolean => {
  const achieved = scoreToTrainingRank(score, kind);
  return meetsTrainingRankByLetter(achieved, requiredRank);
};

File: ios/Jazzify/Training/Engine/TrainingRank.swift
import Foundation

enum TrainingLetterRank: String, Codable, Sendable, CaseIterable {
    case S, A, B, C, D, E, F
}

enum TrainingRank {
    private static let defaultThresholds: [(minScore: Int, rank: TrainingLetterRank)] = [
        (60, .S),
        (50, .A),
        (40, .B),
        (30, .C),
        (20, .D),
        (10, .E),
    ]

    private static let scaleThresholds: [(minScore: Int, rank: TrainingLetterRank)] = [
        (30, .S),
        (25, .A),
        (20, .B),
        (15, .C),
        (10, .D),
        (5, .E),
    ]

    private static let order: [TrainingLetterRank] = [.F, .E, .D, .C, .B, .A, .S]

    private static func thresholds(for kind: TrainingKind) -> [(minScore: Int, rank: TrainingLetterRank)] {
        kind == .scale ? scaleThresholds : defaultThresholds
    }

    static func scoreToRank(_ score: Int, kind: TrainingKind = .chord) -> TrainingLetterRank {
        let normalized = max(0, score)
        for entry in thresholds(for: kind) where normalized >= entry.minScore {
            return entry.rank
        }
        return .F
    }

    static func meetsRank(achieved: TrainingLetterRank, required: TrainingLetterRank) -> Bool {
        guard let achievedIndex = order.firstIndex(of: achieved),
              let requiredIndex = order.firstIndex(of: required)
        else { return false }
        return achievedIndex >= requiredIndex
    }

    static func meetsRequirement(score: Int, requiredRank: TrainingLetterRank, kind: TrainingKind = .chord) -> Bool {
        meetsRank(achieved: scoreToRank(score, kind: kind), required: requiredRank)
    }

    static func parseLetterRank(_ raw: String) -> TrainingLetterRank {
        TrainingLetterRank(rawValue: raw) ?? .F
    }
}
